### PR TITLE
Change back to Guava Cache as Backend

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

This is the first PR of Cache Refactor, contains:
* Change back to Guava Cache as Backend
* Refactor MemoryManager to acquire off-heap memory from Spark
* Generic data access function in FiberCache

Issues need to be resolved in following PR:
* [Important] Keep in-used FiberCache from being evicted
* Change FiberCacheManager, MemoryManger to class for better init
* Decode/Decompress DataFile in off-heap directly
* cache.status may be costly and not considering data eviction

## How was this patch tested?

Unit test passed, also passed new added test suite.

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala [hunk 1]
```diff
@@ -19,36 +19,137 @@ package org.apache.spark.sql.execution.datasources.oap.filecache
...
 
-
 // TODO: make it an alias of MemoryBlock
 trait FiberCache {
-  def fiberData: MemoryBlock
+  // In our design, fiberData should be a internal member.
+  protected def fiberData: MemoryBlock
+
+  // If MemoryManager has no off-heap memory, FiberCache can maintain a Array[Byte] as MemoryBlock
+  def isOffHeap: Boolean = fiberData.getBaseObject == null
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala]: We don't need to expose the API externally right?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 2]
```diff
@@ -57,6 +57,87 @@ private[oap] class CacheResult (
...
+
+  // TODO: Will change to get(), keep it for compatible for now
+  def getOrElseUpdate(fiber: Fiber, conf: Configuration): FiberCache = {
+    val fiberCache = cache.get(fiber, cacheLoader(fiber, conf))
+    // TODO: How to deal with fiber without enough memory. Put into cache or not?
+    if (!fiberCache.isOffHeap) remove(fiber)
+    fiberCache
+  }
+
+  def remove(fiber: Fiber): Unit = {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: I am not sure if we really need to remove the Fiber explicitly, probably not for now. See my other comments, we don't need to evict the data by ourselves, and supposedly, Guava will do that for us if we enable the weight for `cache` and its entries.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala [hunk 3]
```diff
@@ -19,36 +19,137 @@ package org.apache.spark.sql.execution.datasources.oap.filecache
...
+
+  // TODO: a config to control max memory size
+  private val maxMemory = {
+    val sparkMemoryManager = if (SparkEnv.get == null) None else Some(SparkEnv.get.memoryManager)
+    sparkMemoryManager.map { memoryManager =>
+      val oapMaxMemory = (memoryManager.maxOffHeapStorageMemory * 0.7).toLong
+      if (memoryManager.acquireStorageMemory(DUMMY_BLOCK_ID, oapMaxMemory, MemoryMode.OFF_HEAP)) {
+        oapMaxMemory
+      } else {
+        0L
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala]: Better to exit the executor process if we can not get required memory for data caching? YARN or some other resource manager will schedule the executor to other node transparently.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 4]
```diff
@@ -57,6 +57,87 @@ private[oap] class CacheResult (
   val cached: Boolean,
   val buffer: ChunkedByteBuffer)
 
+/** CacheManager
+ *
+ *  TODO: Do we need change object to class for better initialization
+ *  TODO: Change to FiberCacheManger after code is reviewed.
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Just don't want this PR to be too large. I'll update this.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 2 above)
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Maybe we need this for index dropping.

**Comment by linhong-intel** [comment]
@chenghao-intel please take a look at the latest code and let's discuss the next step after that.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 5]
```diff
@@ -45,149 +41,133 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
...
         FiberCacheStatus(dataFile.path, fiberBitSet, fileMeta)
+    }.toSeq
+
+    CacheStatusSerDe.serialize(statusRawData)
+  }
+
+  // TODO: need more discuss about how to put data from file to cache.
+  // Restore back the original implementation first.
+  private def fiber2Data(fiber: Fiber, conf: Configuration): FiberCache = {
+    val data = fiber match {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Can you please add a comment to describe we'd better return the FiberCache directly instead of `Array[Byte]` in the next PR?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 6]
```diff
@@ -45,149 +41,133 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
...
+    Platform.copyMemory(
+      data,
+      Platform.BYTE_ARRAY_OFFSET,
+      memoryBlock.getBaseObject,
+      memoryBlock.getBaseOffset,
+      data.length)
+    DataFiberCache(memoryBlock)
   }
+
+  def getStats: CacheStats = cache.stats()
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: This is great, probably we can introduce a DDL to analysis/print the cache status. Leave it for further another PR.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala [hunk 7]
```diff
@@ -20,35 +20,136 @@ package org.apache.spark.sql.execution.datasources.oap.filecache
...
+   */
+  private val DUMMY_BLOCK_ID = TestBlockId("oap_memory_request_block")
+
+  // TODO: a config to control max memory size
+  private val _maxMemory = {
+    if (SparkEnv.get == null) {
+      throw new OapException("No SparkContext is found")
+    } else {
+      val memoryManager = SparkEnv.get.memoryManager
+      val oapMaxMemory = (memoryManager.maxOffHeapStorageMemory * 0.7).toLong
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala]: Probably we'd better make the `0.7` as configurable.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFile.scala [hunk 8]
```diff
@@ -36,7 +36,7 @@ abstract class DataFile {
   def configuration: Configuration
 
   def createDataFileHandle(): DataFileHandle
-  def getFiberData(groupId: Int, fiberId: Int, conf: Configuration): ChunkedByteBuffer
+  def getFiberData(groupId: Int, fiberId: Int, conf: Configuration): Array[Byte]
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/DataFile.scala]: Returns `MemoryBlock` directly. Or let's leave it for next PR.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/IndexFile.scala [hunk 9]
```diff
@@ -46,7 +46,7 @@ private[oap] trait CommonIndexFile {
  * Read the index file into memory, and can be accessed as [[ChunkedByteBuffer]].
  */
 private[oap] case class IndexFile(file: Path) extends CommonIndexFile {
-  def getIndexFiberData(conf: Configuration): ChunkedByteBuffer = {
+  def getIndexFiberData(conf: Configuration): Array[Byte] = {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/IndexFile.scala]: I will suggest to return the `IndexFiberCache` directly.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataFile.scala [hunk 10]
```diff
@@ -66,7 +66,7 @@ private[oap] case class OapDataFile(path: String, schema: StructType,
     } else dictionaries(fiberId)
   }
 
-  def getFiberData(groupId: Int, fiberId: Int, conf: Configuration): ChunkedByteBuffer = {
+  def getFiberData(groupId: Int, fiberId: Int, conf: Configuration): Array[Byte] = {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataFile.scala]: returns `DataFiberCache`.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala [hunk 11]
```diff
@@ -20,35 +20,136 @@ package org.apache.spark.sql.execution.datasources.oap.filecache
...
+  }
 }
 
 // Data fiber caching, the in-memory representation can be found at [[DataFiberBuilder]]
 case class DataFiberCache(fiberData: MemoryBlock) extends FiberCache
 
+// TODO: This class seems not needed.
+// Because After index partial loading enabled, each MemoryBlock is only for its specific part.
 // Index fiber caching, only used internally by Oap
 private[oap] case class IndexFiberCacheData(
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala]: Change name to `IndexFiberCache`?  As we have the name `DataFiberCache` already.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 12]
```diff
@@ -45,149 +41,133 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
...
-
-    val FiberDataBlock = "fiber_data_(.*)_([0-9]+)_([0-9]+)".r
-    blockId.name match {
-      case FiberDataBlock(fileId, columnIndex, rowGroupId) =>
-        val dataFile = dataFileIdMap(fileId)
-        DataFiber(dataFile, columnIndex.toInt, rowGroupId.toInt)
-      case _ => throw new OapException("unknown blockId: " + blockId.name)
+  private val removalListener = new RemovalListener[Fiber, FiberCache] {
+    override def onRemoval(notification: RemovalNotification[Fiber, FiberCache]): Unit = {
+      logDebug(s"Removing Cache ${notification.getKey}")
```
[the remark this answers is not in the document]
Comment (chenghao-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Another possible crash issue is, the removing FiberCache probably is using by some threads, I am wondering if we have to lock up. We can discuss that offline.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala [hunk 13]
```diff
@@ -20,35 +20,136 @@ package org.apache.spark.sql.execution.datasources.oap.filecache
...
+
+  private def getBaseObj: AnyRef = {
+    // NOTE: A trick here. Since every function need to get memory data has to get here first.
+    // So, here check the if the memory has been freed.
+    if (disposed) throw new OapException("Try to access a freed memory")
+    fiberData.getBaseObject
+  }
+  private def getBaseOffset: Long = fiberData.getBaseOffset
+
+  def getBoolean(offset: Long): Boolean = Platform.getBoolean(getBaseObj, getBaseOffset + offset)
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala]: Add Unit test for those functions.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 14]
```diff
@@ -45,149 +41,133 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
...
+      override def call(): FiberCache = {
+        logDebug(s"Loading Cache $fiber")
+        fiber2Data(fiber, configuration)
+      }
+    }
+  /*
+  private def anotherCacheLoader(fiber: Fiber, configuration: Configuration) = {
+    new Callable[FiberCache] {
+      override def call(): FiberCache = {
+        def calculateFiberSize(fiber: Fiber): Int = {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Yes, we need to think about this also...let's see how to implement it.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 15]
```diff
@@ -17,39 +17,32 @@
...
   // for any FiberData, the first defaultSize / 8 will be the bitmask
   // TODO what if defaultSize / 8 is not an integer?
 
   // TODO get the bitset from the FiberByteData
   val bitset: BitSet = {
     val bs = new BitSet(defaultSize)
     val longs = bs.toLongArray()
-    Platform.copyMemory(baseObject, baseOffset,
-      longs, Platform.LONG_ARRAY_OFFSET, longs.length * 8)
+    buffer.copyMemory(0, longs, Platform.LONG_ARRAY_OFFSET, longs.length * 8)
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: Sure

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 16]
```diff
@@ -137,8 +131,7 @@ class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: ChunkedByte
     val length = getIntValue(idx * 2)
     val offset = getIntValue(idx * 2 + 1)
     val result = new Array[Byte](length)
-    Platform.copyMemory(baseObject, baseOffset + offset, result,
-      Platform.BYTE_ARRAY_OFFSET, length)
+    buffer.copyMemory(offset, result, Platform.BYTE_ARRAY_OFFSET, length)
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: Sure

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 17]
```diff
@@ -45,149 +41,133 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
...
-  }
-}
-
-private[oap] class CacheResult (
-  val cached: Boolean,
-  val buffer: ChunkedByteBuffer)
-
 /**
- * Fiber Cache Manager
+ * CacheManager
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: np

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 18]
```diff
@@ -45,149 +41,133 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
...
-        // So, use on-heap memory after failure.
-        if (blockManager.putBytes(blockId, offHeapBytes, storageLevel)) {
-          logDebug("Put fiber to cache success")
-          new CacheResult(true, blockManager.getLocalBytes(blockId).get)
-        } else {
-          logDebug("Put fiber to cache fail")
-          offHeapBytes.dispose()
-          new CacheResult(false, bytes)
+  private val MB: Double = 1024 * 1024
+  private val maxWeight = (MemoryManager.maxMemory / MB).toInt
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: np

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala (hunk 13 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala]: done

**Comment by linhong-intel** [comment]
@chenghao-intel @adrian-wang 
Please review this PR and maybe we can merge it after that.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 19]
```diff
@@ -45,149 +39,85 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
...
+      .maximumWeight(MAX_WEIGHT)
+      .weigher(weigher)
+      .build[Fiber, FiberCache]()
+
+  def get(fiber: Fiber, conf: Configuration): FiberCache = {
+    // Used a flag called disposed in FiberCache to indicate if this FiberCache is removed
+    cache.get(fiber, cacheLoader(fiber, conf))
   }
 
-  def status: String = {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Basically there will be 2 issues:
1) Data eviction happens when we calculate the status;
2) Performance is very bad as we have to get the Cache Handle first, which probably very costly.

let's think about it in next PR.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 20]
```diff
@@ -45,149 +39,85 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
...
-    }
+    }.toSeq
+
+    CacheStatusSerDe.serialize(statusRawData)
+  }
 
-    val retStatus = CacheStatusSerDe.serialize(statusRawData)
-    retStatus
+  private def fiber2Data(fiber: Fiber, conf: Configuration): FiberCache = {
+    fiber match {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Can we add the API `getFiberData` in the `Fiber` trait? and then we don't need the pattern match here.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 19 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: added these issues in PR description and TODO in code. I'll address in next PR.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexFileReader.scala [hunk 21]
```diff
@@ -51,23 +52,14 @@ private[oap] case class BTreeIndexFileReader(
   private def getIntFromBuffer(buffer: Array[Byte], offset: Int) =
     Platform.getInt(buffer, Platform.BYTE_ARRAY_OFFSET + offset)
 
-  def readFooter(): Array[Byte] = {
-    val footerBuffer = new Array[Byte](footerLength)
-    reader.readFully(footerIndex, footerBuffer)
-    footerBuffer
-  }
+  def readFooter(): FiberCache =
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexFileReader.scala]: Is there any case that the FiberCache is invalid (evicted maybe) in later use as the footer(or rowIds) is accessed all through the query?

Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexFileReader.scala]: Oh, the return value is a cache entry index, I think you may change FiberCache => FiberCacheEntry.

Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexFileReader.scala]: Thanks. This is really a problem.

Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexFileReader.scala]: But this should be same to the lock issue mentioned in PR description. So this will be fixed in next PR.

**Comment by linhong-intel** [comment]
merge it and let's fix issues in the following PR
